resurrect the deb_packages table

maclennann opened the pull request:
This is an attempt to pick up where #324 / #325 left off.

I basically resurrected everything as it stood from that PR with minor changes to get it compiling again. The only notable change is `:145` and `:146` - without closing the package DB between queries, `pkg_array_init_from_db` bombs out when it tries to load in each of the packages again thinking they're duplicates.

I wasn't sure if there was any technical reason that #325 was closed, or if it was just inactivity. If there was any further work to be done for it, I'd be happy to take a crack.

maclennann commented:
Cool, thanks for the quick review and response.

I smoothed over the `dpkg` version incompatibilities. I don't know enough c++ to know if there was a more correct or idiomatic way of doing it, but the missing functions were generally just small convenience functions around other parts of the public API, so I just brought the same functionality forward.

I can give you `md5sum` and `revision` columns so the tables have a similar structure, but there is no good way to put any meaningful data in either one.

`dpkg` [doesn't have the same notion of package-level verification](http://serverfault.com/a/322554) as `rpm` has. `dpkg` has md5s of a subset of files in `/var/lib/dpkg/info/{package}.md5sums` and those are what it (and `debsums`) verifies against. There is no single, package-wide md5sum.

`revision` was, historically, the version of the package (as tracked separately from the version of the application contained within the package). This field has since been [obsoleted](http://anonscm.debian.org/cgit/dpkg/dpkg.git/tree/lib/dpkg/fields.c?h=1.17.x#n251) and [rolled right into the version](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version).

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+      continue;
+    }
+
+    results.push_back(extractDebPackageInfo(pkg, &pkg->installed));
```
Why pass `pkg` and `&pkg->installed` here?

Also, you can pass `&results` as a param and avoid copying the returned Row, like: https://github.com/facebook/osquery/blob/master/osquery/tables/system/crontab.cpp#L58 just a nit.

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+
+QueryData genDebs(QueryContext &context) {
+  QueryData results;
+  struct pkg_array array;
```
nit: array might be a confusing name

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+* Fields information - taken from lib/dpkg/parse.c
+*/
+const struct fieldinfo fieldinfos[] = {
+    /* Note: Capitalization of field name strings is important. */
```
Can we change these to C++ style comments. The doxygen generation comment blocks above methods/data structures are C style though. :) 

Also, just curious, do you need to enumerate all the fields, can we limit to just the fields that are emitted in the table spec? Or do you recommend keeping them all here for completeness/future additions.

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+    {"Description", f_charfield, w_charfield, PKGIFPOFF(description)},
+    {"Triggers-Pending", f_trigpend, w_trigpend},
+    {"Triggers-Awaited", f_trigaw, w_trigaw},
+    /* Note that aliases are added to the nicknames table. */
```
Same comment nit above

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+  want_hold,
+  want_deinstall,
+  want_purge,
+  /** Not allowed except as special sentinel value in some places. */
```
nit, change to C++ style comments

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+  /** Not allowed except as special sentinel value in some places. */
+  want_sentinel,
+} want;
+/** The error flag bitmask. */
```
nit, change to C++ style comments

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+  return strcmp(arch_a, arch_b);
+}
+
+void dpkg_setup(char const *name,
```
Do we need to pass a `const *` if this isn't used anywhere but as an aux internal setup helper? Can we call `dpkg_set_progname` directly `dpkg_set_progname("osquery");`? And nit about the potentially confusing `array` variable name. ;)

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+  modstatdb_open(mode);
+
+  pkg_array_init_from_db(array);
+  pkg_array_sort(array, pkg_sorter);
```
Sweet workaround to support both 12/14!

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+  pkg_array_sort(array, pkg_sorter);
+}
+
+void dpkg_teardown(struct pkg_array *array) {
```
nit, potentially confusing `array` variable name.

theopolis commented on osquery/tables/system/linux/deb_packages.cpp:
```diff
@@ -0,0 +1,185 @@
+
+/**
+* Fields information - taken from lib/dpkg/parse.c
+*/
```
Should be written:
```cpp
/**
 * @brief Fields information - taken from lib/dpkg/parse.c
 */
```

theopolis commented:
Nice, let me know if my recent comments make sense. If you don't mind adding those last few fixups we can merge and iterate on the revision/hashing a bit later.

It would be great to have an integrity check of the package-installed contents somehow/someway. And we can later edit revision to be a duplicate of the version. That'll be helpful when comparing results from rpm/deb from a splunk or other log aggregation search/schedule interface.

maclennann commented:
Thanks. I'm pretty sure I addressed all of your comments.

I found that dpkg does track the revision as a discrete part of the version, but just doesn't provide a function to extract it when parsing the package. So I made one and added the field it to the table.

I can put the various filenames/md5s into an md5 column, but having multiple values in a single row like that doesn't seem right. I could do something like taking all of the md5s of all the files and calculating an aggregate md5 of them just to have some data to work against, but it certainly wouldn't be the dpkg-official way of handling things and might get confusing to users since it isn't a "real" thing.

theopolis commented:
Awesome! Let's table the hashing to a RFC issue. I'm thinking an aggregate md5 of the files provided by the package is useful but intense. If we do include an integrity check it should be useful and stable as we iterate the table (not so important that we diverge from Debian package standards/provides).
